### PR TITLE
Implementing PUT routine

### PR DESCRIPTION
**Pull request by ipdemes**
Implementing PUT method 
The implementation reuses `WRAP` task (with some modifications)  to compute points to the source array that need to be updated with the Values.
Since Legate Copy operation modifies values asynchronously,  we can't guarantee that the output array will be the same as output from numpy.put for the case when there are duplicates in indices. Therefore, I have added the check for duplicate indices and warning message if they are found.

### CODE REVIEW

File: cunumeric/array.py
Comment (ipdemes) [inline cunumeric/array.py:2516]: In general, this warning can be applied to any case when there are duplicated indices. should I add the check for it?
something like:
```
indices_tmp = np.unique (indices)
if indices_tmp.size < indices.size:
    runtime.warn(
```

File: cunumeric/array.py
Comment (bryevdv) [inline cunumeric/array.py:2489]: This would match numpy:

```suggestion
            raise ValueError(f"clipmode must be one of 'clip', 'raise', or 'wrap' (got {mode})")

")
```

File: cunumeric/array.py
Comment (bryevdv) [inline cunumeric/array.py:2511]:
```suggestion
                " might yield results different from NumPy",
```

File: cunumeric/array.py
Comment (bryevdv) [inline cunumeric/array.py:2500]: I feel like this happens somewhat frequently. Is it worth having a *"check dtype and covert with a warning if necessary"* helper function?

**Review by bryevdv — APPROVED** [review]
LGTM modulo the bad copy/merge that needs fixed in `array.py`

File: cunumeric/eager.py
Comment (manopapad) [inline cunumeric/eager.py:624]: Add this check above:
```
self.check_eager_args(indices, values)
```

File: cunumeric/deferred.py
Comment (manopapad) [inline cunumeric/deferred.py:790]: Why do you need to do this? It looks like you're overriding the shape of the target array if it's backed by a future. Did you see a case where a future-backed array had the wrong shape?

File: cunumeric/array.py
Comment (manopapad) [inline cunumeric/array.py:3857]: There's a couple more places in array.py and module.py where this pattern appears. Could you also convert those to use `_warn_and_convert`?

File: cunumeric/deferred.py
Comment (manopapad) [inline cunumeric/deferred.py:1671]: I don't think you need to do that, since `indices` itself is not used in a copy

File: cunumeric/deferred.py
Comment (manopapad) [inline cunumeric/deferred.py:1676]: You are creating a copy of `self` here and writing to that. This has the effect that any values written are not reflected on the original array:

```
x = cn.zeros(1)
cn.put(x, cn.arange(4), cn.ones(4), mode="clip")
print(x) # prints [0]
```

Possibly you want to do something like what you do in `set_item` https://github.com/nv-legate/cunumeric/blob/branch-22.10/cunumeric/deferred.py#L909-L912.

File: cunumeric/module.py
Comment (manopapad) [inline cunumeric/module.py:3454]: I find this a little cryptic. It also doesn't make explicit that the indices act as if the `a` array was flattened. Could we do something like:

```
Replaces specified elements of an array with given values.

The indexing works as if the target array had been flattened.
```

File: cunumeric/module.py
Comment (manopapad) [inline cunumeric/module.py:3463]: The original documentation also states that "If v is shorter than ind it will be repeated as necessary." We should mention this here as well (assuming we are keeping this behavior).

File: cunumeric/array.py
Comment (manopapad) [inline cunumeric/array.py:2469]: I would redirect readers to `cunumeric.put`:

```
        """
        Replaces specified elements of the array with given values.

        Refer to :func:`cunumeric.put` for full documentation.

        See Also
        --------
        cunumeric.put : equivalent function

        Availability
        --------
        Multiple GPUs, Multiple CPUs
        """
```

File: cunumeric/array.py
Comment (manopapad) [inline cunumeric/array.py:2482]: just `mode` rather than `clipmode`

File: cunumeric/array.py
Comment (manopapad) [inline cunumeric/array.py:2483]: `got {mode}`

File: cunumeric/module.py
Comment (ipdemes) [reply cunumeric/module.py:3463]: Thank you, I will add this. Yes. we are keeping this behavior

File: cunumeric/deferred.py
Comment (ipdemes) [reply cunumeric/deferred.py:790]: Yes, for the case when scalar is converted to the ndarray it will get `()` shape. But we actually want it to be `(1,)` shape

File: cunumeric/array.py
Comment (ipdemes) [reply cunumeric/array.py:3857]: sure

File: cunumeric/deferred.py
Comment (ipdemes) [reply cunumeric/deferred.py:1671]: I believe I need to do that when indices is a scalar that was converted to ndarray, e.g it has shape `()`. See above

File: cunumeric/deferred.py
Comment (ipdemes) [reply cunumeric/deferred.py:1676]: you are right. Thank you!

File: cunumeric/array.py
Comment (manopapad) [inline cunumeric/array.py:2505]: This is an expensive check, and I wouldn't like to have it happen indiscriminately. I think it's sufficient to note this on the documentation for `put`.

File: cunumeric/deferred.py
Comment (manopapad) [reply cunumeric/deferred.py:1679]: Now the same example fails for me with
```
Traceback (most recent call last):
  File "/Users/mpapadakis/legion/bindings/python/build/lib/legion_top.py", line 440, in legion_python_main
    run_path(args[start], run_name='__main__')
  File "/Users/mpapadakis/legion/bindings/python/build/lib/legion_top.py", line 255, in run_path
    exec(code, module.__dict__, module.__dict__)
  File "a.py", line 10, in <module>
    cn.put(x, cn.arange(4), cn.ones(4), mode="clip")
  File "/Users/mpapadakis/legate.core/legate/core/context.py", line 232, in wrapper
    result = func(*args, **kwargs)
  File "/Users/mpapadakis/cunumeric/cunumeric/coverage.py", line 113, in wrapper
    return func(*args, **kwargs)
  File "/Users/mpapadakis/cunumeric/cunumeric/array.py", line 143, in wrapper
    return func(*args, **kwargs)
  File "/Users/mpapadakis/cunumeric/cunumeric/module.py", line 3479, in put
    a.put(indices=indices, values=values, mode=mode)
  File "/Users/mpapadakis/legate.core/legate/core/context.py", line 232, in wrapper
    result = func(*args, **kwargs)
  File "/Users/mpapadakis/cunumeric/cunumeric/coverage.py", line 113, in wrapper
    return func(*args, **kwargs)
  File "/Users/mpapadakis/cunumeric/cunumeric/array.py", line 143, in wrapper
    return func(*args, **kwargs)
  File "/Users/mpapadakis/cunumeric/cunumeric/array.py", line 2510, in put
    self._thunk.put(indices._thunk, values._thunk)
  File "/Users/mpapadakis/cunumeric/cunumeric/deferred.py", line 117, in wrapper
    return func(*args, **kwargs)
  File "/Users/mpapadakis/cunumeric/cunumeric/deferred.py", line 1687, in put
    N = self.ndim
AttributeError: 'NoneType' object has no attribute 'ndim'
```

File: cunumeric/array.py
Comment (ipdemes) [reply cunumeric/array.py:2505]: Sure, we can do that. I was also thinking of checking this only in the debug mode, but, for this, we will have to have a `--debug` flag implemented for the python (or something like this)

File: cunumeric/array.py
Comment (manopapad) [reply cunumeric/array.py:3858]: On second thought, most of those places were not emitting warnings before, and I don't want to reason in this PR about whether each of those places should have warnings or not, so I just pushed a change to just use `astype` where the original code was not already throwing a warning.

Comment (ipdemes) [reply cunumeric/array.py:3858]: Thank you for doing that. I was expecting follow-up on this after I pushed the changes

File: cunumeric/deferred.py
Comment (ipdemes) [reply cunumeric/deferred.py:1679]: Should be fixed now. I should have checked your example earlier

File: src/cunumeric/index/wrap_template.inl
Comment (manopapad) [inline src/cunumeric/index/wrap_template.inl:88]: Can the two branches be combined in a single declaration?
```
WrapArgs args{context.outputs()[0], shape, has_input, has_input ? context.inputs()[0] : Array()};
```

File: src/cunumeric/index/wrap_template.inl
Comment (manopapad) [inline src/cunumeric/index/wrap_template.inl:71]: It doesn't matter whether the `in` store is dense, since you always access that through the accessor interface (which works whether dense or not). It only matters for `out`, which you access directly through the pointer (if dense).

File: src/cunumeric/index/wrap.cu
Comment (manopapad) [inline src/cunumeric/index/wrap.cu:36]: Can we pass the `IND` argument with `const`, just to match the other cases? I.e. `const IND indices` for kernels, and `const IND& indices` on host-side functions (in this and other files?)

It might just be me, but anytime I see a non-const `Foo&` argument, that signals to me that this is an "output" argument (which is not the case here AFAICT).

File: cunumeric/array.py
Comment (manopapad) [reply cunumeric/array.py:2505]: Yes, let's discuss this at the next meeting. We already have two "debugging mode" flags (`LEGATE_TEST` and `-cunumeric:test`), and we're talking about introducing another one, so I'd like to get everyone's opinion before we go and do this.

For now let's just move this warning to the documentation on `cunumeric.put`.

Comment (ipdemes) [reply cunumeric/array.py:2505]: sounds good

File: src/cunumeric/index/wrap_template.inl
Comment (ipdemes) [reply src/cunumeric/index/wrap_template.inl:88]: sure, thank you

File: src/cunumeric/index/wrap_template.inl
Comment (ipdemes) [reply src/cunumeric/index/wrap_template.inl:71]: right.

File: src/cunumeric/index/wrap.cu
Comment (ipdemes) [reply src/cunumeric/index/wrap.cu:36]: sure, thanks

File: cunumeric/array.py
Comment (manopapad) [inline cunumeric/array.py:2490]: I don't think we currently do any checks in the case where `mode == "raise"`.

Comment (ipdemes) [reply cunumeric/array.py:2490]: that is right, just pushed the check for CPU and OpenMP. To my understanding we are not allowed to throw exceptions in CUDA kernels, is that correct?

File: cunumeric/deferred.py
Comment (ipdemes) [inline cunumeric/deferred.py:3398]: @manopapad : I had to explicitly convert src to deferred array in the case when src in a scalar array . I thought auto_convert[1] should have taken care of this. Any thoughts ?

File: cunumeric/array.py
Comment (manopapad) [reply cunumeric/array.py:2490]: Correct, you would need to have the kernels signal this failure in a different way, e.g. reducing into a shared boolean variable that an out-of-bounds index was found. Or run a second kernel that looks for such indices.

But at that point it may be cleaner to not do this check inside the wrap task, instead add a 3rd case to `array.py`:
```
else:  # mode == "raise"
  if indices.min() < 0 or indices.max() >= self.size:
    raise ...
```
This way you also skip the check in case the user passes "wrap" or "clip" (the check is unnecessary in that case).

Comment (manopapad) [reply cunumeric/array.py:2490]: Now that you mention it, wouldn't this also apply to the zip kernel? Isn't it also bad that `compute_idx` throws an exception there as well?

Comment (ipdemes) [reply cunumeric/array.py:2490]: That makes sense. 
However, the down side of having the check on the python side is that it will invoke 2 more tasks: `min` and `max`. So,  I think it would be better to make CUDA task to return error . What do you think?
And yes, the same issue applies to the ZIP kernel. In that case compiler didn't complain about exception because it is hidden in the constexpr function

Comment (manopapad) [reply cunumeric/array.py:2490]: Yes, good point. The simplest solution might be to run a second kernel (in the wrap and zip tasks) to look for out-of-bounds indices - `thrust::minmax_element` might be a good choice. And then don't do the bounds checks in the "main" wrap/zip kernel (you would need to have a separate `compute_idx` function I assume for .cu files).

File: cunumeric/deferred.py
Comment (manopapad) [reply cunumeric/deferred.py:790]: That makes sense, however I would it to be more explicit. Can we just check in `_convert_future_to_regionfield` whether `self.shape == ()`, and in that case do `shape = (1,)`? That would make the purpose of this obvious. In that case I think you don't need the `future: bool` argument.

Comment (ipdemes) [reply cunumeric/deferred.py:790]: yes, this is rather a special case for the `wrap`. I will look how to change the logic in wrap so that we don't have to add this change to  `_convert_future_to_regionfield`

File: cunumeric/deferred.py
Comment (manopapad) [reply cunumeric/deferred.py:3398]: https://github.com/nv-legate/cunumeric/pull/640 fixes this (you don't need to call `to_deferred_array` manually). If you don't want to wait for that, then you can pass the arguments as follows in `array.py::_wrap`:

```
        out._thunk._wrap(self._thunk, new_len=new_len)
```

**Comment by manopapad** [comment]
I am seeing the following error when running these examples:

```
cn.put(cn.zeros(()), [0], 1)
cn.put(cn.zeros(()), [0], [1])
```

```
  File "/Users/mpapadakis/legate.core/legate/core/operation.py", line 164, in add_alignment
    raise ValueError(
ValueError: Stores must have the same shape to be aligned, but got Shape(()) and Shape((1,))
```

I also see the same error on this example, if I also merge https://github.com/nv-legate/cunumeric/pull/530:

```
cn.put(cn.zeros(()), 0, 1)
```

File: cunumeric/deferred.py
Comment (ipdemes) [reply cunumeric/deferred.py:3398]: Thank you, Manolis!

**Comment by ipdemes** [comment]
@manopapad : thank you for finding the error. I am working on the fix

File: cunumeric/deferred.py
Comment (ipdemes) [reply cunumeric/deferred.py:790]: @manolis, actually I don't know how to do this without converting the shape of the futures for put routine. We need to do this not for the wrap task, but for the copy operation after the task. I can convert shape of the indirect field only in the task, but not the shape of indices or values there.
how about renaming this boolean variable to "change_shape" instead of the "future" in the `_convert_future_to_regionfield`?

Comment (manopapad) [reply cunumeric/deferred.py]: That's fine, I think it would be cleanest to check in `_convert_future_to_regionfield` something like:
```
if change_shape and shape == ():
  shape = (1,)
```
and pass `change_shape=True` whenever this is relevant. Then it's obvious that we're only changing `()` to `(1,)`, and not other shapes.

Comment (ipdemes) [reply cunumeric/deferred.py]: Sounds good! thank you

**Comment by ipdemes** [comment]
> @manopapad : thank you for finding the error. I am working on the fix

@manopapad : the fix for this use case is blocked on https://github.com/nv-legate/cunumeric/pull/582#issuecomment-1268951364 issue

**Comment by ipdemes** [comment]
@manopapad I think I have addressed all the comments/issues with my latest commits.

File: cunumeric/deferred.py
Comment (ipdemes) [inline cunumeric/deferred.py:2946]: I have added this logic for the case when `lhs` has shape `()` and `rhs` has shape `(1,)`. Otherwise we get an error

File: cunumeric/deferred.py
Comment (ipdemes) [reply cunumeric/deferred.py:3403]: I will wait until #640 is merged

File: cunumeric/array.py
Comment (manopapad) [inline cunumeric/array.py:2513]: You need to check here that all values in `indices` are zero; otherwise you would miss an error like:
```
>>> x = np.zeros(())
>>> x
array(0.)
>>> np.put(x, [1], [13])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<__array_function__ internals>", line 180, in put
  File "/Users/mpapadakis/opt/miniconda3/envs/legate/lib/python3.8/site-packages/numpy/core/fromnumeric.py", line 543, in put
    return put(ind, v, mode=mode)
IndexError: index 1 is out of bounds for axis 0 with size 1
```

File: cunumeric/deferred.py
Comment (manopapad) [reply cunumeric/deferred.py]: I merged the fix and updated the code accordingly.

File: src/cunumeric/index/zip.cu
Comment (manopapad) [inline src/cunumeric/index/zip.cu:129]: The kernel is reducing to a value of type `bool`, so it should be `sizeof(bool)` here.

File: src/cunumeric/index/zip.cu
Comment (manopapad) [inline src/cunumeric/index/zip.cu:108]: Isn't the second condition `index >= extent`?

File: cunumeric/array.py
Comment (ipdemes) [reply cunumeric/array.py:2514]: makes sense, thank you

File: src/cunumeric/index/advanced_indexing.cu
Comment (manopapad) [inline src/cunumeric/index/advanced_indexing.cu:97]: Here the reduction accumulator is of type `uint64_t`, so we need to allocate 8 bytes per warp. The type used in the `sizeof` was technically incorrect (`int64_t`, it should be `uint64_t`), but I already handled this in https://github.com/nv-legate/cunumeric/pull/659, so no changes are needed to this file.

File: cunumeric/array.py
Comment (manopapad) [inline cunumeric/array.py:2512]: I think we should also allow `-1`, i.e. it should be something like this?
```
if indices.min() < -1 or indices.max > 0:
```

File: src/cunumeric/index/wrap.cu
Comment (manopapad) [inline src/cunumeric/index/wrap.cu:100]: `sizeof(bool)`

File: cunumeric/deferred.py
Comment (manopapad) [reply cunumeric/deferred.py:2965]: Is this still required? I just ran the test suite without this fix and everything seems to be working.

File: src/cunumeric/index/advanced_indexing.cu
Comment (ipdemes) [reply src/cunumeric/index/advanced_indexing.cu:97]: Thank you, Manolis!

File: src/cunumeric/index/wrap.cu
Comment (ipdemes) [reply src/cunumeric/index/wrap.cu:100]: right

**Comment by manopapad** [comment]
This is ready to be merged, except for one remaining question above. I am noting here some work we should address in future PRs:

(1) We can skip the out-of-bounds checks inside of the task if `mode != "raise"`.

(2) There are currently two different values in the "wrap" files that are both confusingly referred to as "input"/"in". I would like us to rename all values referring to the target array as "tgt" or "base", to make the code easier to understand. E.g. `input_rect` in `wrap_template.inl` would become `base_rect`, `pitches_in` in `wrap.cc` would become `pitches_base` etc.

(3) Raising an exception within an OpenMP block causes that exception to be caught and reported at the end of that block, instead of propagating it to our catch statement https://github.com/nv-legate/legate.core/blob/branch-22.12/src/core/task/task.h#L111. This causes code like the following to report an exception inside C++ instead of propagating that exception to python:

```
x = cn.zeros((10,))
cn.put(x, [6,7,8,9,10], 1)
```

Note that this likely affects `zip` as well.

File: cunumeric/deferred.py
Comment (ipdemes) [reply cunumeric/deferred.py:2965]: I don't think it is tested. I will add a test

**Comment by ipdemes** [comment]
@magnatelee : thank you, I have created an issue for the remaining improvements/fixes : https://jirasw.nvidia.com/browse/CNM-438

**Review by manopapad — APPROVED** [review]

**Pull request merged**
